### PR TITLE
[FEAT][kernels] Add fused ratio-clip-aggregate loss primitive

### LINKED ISSUE #252: [FEAT][kernels] Fused ratio-clip-aggregate loss primitive for PPO/GRPO
## Motivation

The roadmap tracks both `fused_ppo_clip_loss` and a generalized ratio-clip-aggregate policy-gradient primitive. Today, `ratio_kl` (#40 / PR #97) efficiently produces per-token ratios and KL terms, but downstream PPO/GRPO clipping, masking, reductions, and metrics are still composed from multiple PyTorch operations in `TritonGRPOLossOp` and the DeepSpeed trainer.

This creates extra full-size temporaries and kernel launches on the policy-loss hot path.

## Proposed operator

Add a dispatchable `ratio_clip_aggregate` operator with PyTorch reference and Triton implementations.

Inputs:

- per-token policy ratio
- detached advantages, either per-token `[B, T]` or per-sequence `[B]`
- active-token mask `[B, T]`
- asymmetric `clip_low` / `clip_high`
- optional per-token penalty terms and coefficient (for the existing KL path)

Outputs:

- total loss
- clipped policy loss
- optional mean penalty
- clip fraction

## Semantics and constraints

- Match `-min(ratio * advantage, clamp(ratio, 1-clip_low, 1+clip_high) * advantage)` over active tokens.
- Avoid materializing broadcast advantages or per-token surrogate tensors.
- Preserve gradients to `ratio` and optional penalty terms; advantages are detached RL targets.
- Use deterministic staged reductions without `atomicAdd` and without CPU-GPU synchronization.
- Define safe empty-mask behavior.
- Keep the interface extensible for PPO/GRPO first and later asymmetric objective configurations without claiming GSPO/DAPO support in v1.

## Planned PR scope

1. Native PyTorch reference and public behavior tests.
2. Triton forward/backward kernels and registry dispatch.
3. Integrate the primitive into `TritonGRPOLossOp` (reusing `ratio_kl`, not duplicating its online-softmax work).
4. Correctness, gradient, masking, BF16/FP16/FP32, and batch-shape coverage.
5. Focused latency/VRAM benchmark against the current PyTorch composition.
6. Operator documentation and explicit fallback behavior.

Roadmap: #83
Related foundation: #40, #46, PR #97

### PR DESCRIPTION
**Pull request by Chen-BUPT**
Closes #252.

## What changed

- Add native PyTorch and Triton `ratio_clip_aggregate` backends for PPO/GRPO clipped-surrogate loss.
- Fuse asymmetric ratio clipping, active-token masking, optional penalty reduction, clip-fraction reporting, and analytical backward.
- Accept per-sequence advantages directly, avoiding a broadcast `[B, T]` tensor.
- Use a tuned single-program reduction through 65,536 tokens and a deterministic two-stage reduction for larger inputs; neither path uses `atomicAdd` or CPU synchronization.
- Register CUDA, ROCm, and CPU backends and integrate the operator into GRPO and the DeepSpeed training worker.
- Add correctness, gradient, dtype, non-contiguous input, empty-mask, reduction-boundary, determinism, registry, benchmark, and operator documentation coverage.

## Performance

Measured on NVIDIA RTX PRO 5000 Blackwell, PyTorch 2.11, CUDA 12.9, Triton 3.6, FP32 inputs, 90% mask density, 100 warmups, and 500 iterations.

| Shape | PyTorch fwd | Triton fwd | Speedup | PyTorch fwd+bwd | Triton fwd+bwd | Speedup |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| 32 x 256 | 0.143 ms | 0.036 ms | 4.02x | 0.416 ms | 0.178 ms | 2.34x |
| 128 x 1024 | 0.153 ms | 0.050 ms | 3.04x | 0.429 ms | 0.201 ms | 2.14x |
| 256 x 4096 | 0.154 ms | 0.050 ms | 3.05x | 0.432 ms | 0.200 ms | 2.16x |
| 256 x 16384 | 0.317 ms | 0.051 ms | 6.28x | 0.664 ms | 0.200 ms | 3.32x |

At 4.19M tokens, measured forward intermediates decrease from 64.0 MiB to 0.3 MiB.

## Validation

- `pytest -q tests/test_ratio_clip_aggregate.py tests/test_grpo_loss.py tests/test_deepspeed_training_worker.py` — 58 passed
- `pytest -q tests/test_kernel_registry.py tests/test_rl_kernel_loss_step.py tests/test_ratio_kl.py tests/test_grpo_single_gpu_example.py` — 29 passed, 1 skipped
- `ruff check` and `ruff format --check` on every changed Python file
- `mkdocs build --strict -f mkdocs.yaml`
- `python benchmarks/benchmark_ratio_clip_aggregate.py --iterations 500 --warmup 100`

The disposable full-suite container did not include the repository's deterministic-attention CUDA extension, so its extension-dependent tests were not a valid signal; the affected operator and integration suites above are green.

## Summary by CodeRabbit

* **New Features**
  * Added ratio-clip aggregation for policy loss, masking, advantage weighting, optional penalties, KL reduction, and clip-fraction reporting.
  * Added CPU, CUDA, and ROCm support with automatic backend selection.
  * Integrated the operator into GRPO training workflows.

* **Documentation**
  * Added operator documentation covering usage, behavior, limitations, and performance.
  * Updated GRPO loss documentation and navigation.

* **Tests**
  * Added comprehensive coverage for validation, gradients, reductions, dispatch, and edge cases.

* **Benchmarks**
  * Added configurable performance and memory benchmarks for native and accelerated implementations.

### CODE REVIEW

**Review by Flink-ddd — APPROVED** [review]
LGTM! Thanks for the contribution.

**Comment by Flink-ddd** [comment]
cc @KJLdefeated PTAL, Thank you.
